feat(Button): extended inline styling

thatblindgeye opened the pull request:
Closes #5351 

@mceledonia @mcoker This draft currently only extends inline support to the plain button. I did try seeing what similar styling might look like for other button variants, for example the primary variant:

![Inline primary example](https://github.com/patternfly/patternfly/assets/70952936/a1665d63-776e-4e2a-9dfc-d10f61274fc1)

One thing to note is that with the Tabs component's action buttons (close and help), the icon becomes a little bigger than it currently is (can be fixed by adding a font size CSS var for pf-v5-c-tabs__item-action), and the outline on focus gets a little overlapped on the top and bottom:

![Tabs action button outline](https://github.com/patternfly/patternfly/assets/70952936/8568db0a-011b-4d99-b449-fa51e88ea2d5)


 We could use the current outline offset in the Tabs css (-0.1875rem), though that might make it feel cramped using `--pf-v5-global--FontSize--xs` (worse if the font size is custom and smaller than this):

![Tabs action buttons outline with offset](https://github.com/patternfly/patternfly/assets/70952936/3d8d3ada-a2e9-4ae5-8bf1-87e6045e3e6c)

This is also assuming we'd want to move the button styling out of the Tabs scss. If not, then it's a bit simpler to keep the `pf-v5-c-tabs__item-action .pf-v5-c-button` block and add the outline offset there.

srambach commented on src/patternfly/components/Button/button.scss:
```diff
@@ -553,6 +559,12 @@
     background-color: var(--#{$button}--m-plain--BackgroundColor);
 
+    &.pf-m-inline {
+      padding: var(--#{$button}--m-plain--m-inline--PaddingTop) var(--#{$button}--m-plain--m-inline--PaddingRight) var(--#{$button}--m-plain--m-inline--PaddingBottom) var(--#{$button}--m-plain--m-inline--PaddingLeft);
```
Do you want to go ahead and make this logical properties?

thatblindgeye replied:
Do we plan on updating CSS vars similarly at any point, e.g. instead of `--component--PaddingTop` => `--component--PaddingBlockStart` or something along those lines? Probably makes more sense to keep the `PaddingTop` etc var names to be less confusing, but curious if that's on the roadmap in regards to RTL.

srambach replied:
@coker and I talked about that - we were thinking to keep the variables as is because that would be a sweeping breaking change.

mcoker commented on src/patternfly/components/Button/button.scss:
```diff
@@ -143,6 +143,12 @@
   --#{$button}--m-plain--active--Color: var(--#{$pf-global}--Color--100);
   --#{$button}--m-plain--disabled--Color: var(--#{$pf-global}--disabled-color--200);
   --#{$button}--m-plain--disabled--BackgroundColor: transparent;
+  --#{$button}--m-plain--m-inline--FontSize: inherit;
```
`inherit` (and `unset` and `initial`) can't be used as the value of a CSS var, and have that value come out of the `var()` function that refers to it. You'll need to use `inherit` as the fallback to a CSS var in the `var()` function. From there you can trigger the fallback by setting the var's value to `initial`. Technically `inherit` will trigger the fallback if one is set, too, but `initial` is the designated value to always trigger the fallback. 
You can read about that here - https://www.w3.org/TR/css-variables-1/#example-ee865e1b

Your PR is working because in `font-size: var(--#{$button}--m-plain--m-inline--FontSize);`, the custom property in the `var()` function is invalid, so `var()` is returning `initial` as the value... and the `initial` value for `font-size` is `inherit`.

mcoker reviewed (commented):
Nice! I left some comments, but I wonder if we should look at this overall. I think there are probably 2 use cases here:
1) Use for a small/compact plain button outlined in https://github.com/patternfly/patternfly/issues/5351
2) An "inline" button that adapts its size to the context where it's used. 

I think this is more along the lines of an inline button, but the left/right padding set to "sm" doesn't seem ideal as the text scales. If we wanted the padding to scale, it could be based off of the font-size (`1em`). Though if we followed the non-inline plain button as a guide for that padding, its padding is `16px`, so the left/right would just be `1em`, which is too big as the button icon size scales down.

There is a comment in the original issue around discussing a product need for this https://github.com/patternfly/patternfly/issues/5351#issuecomment-1493133772

@nicolethoen do you recall that conversation? I looked in the office hours agenda before that comment and skimmed a couple of the recordings and didn't see that discussion.

thatblindgeye commented on src/patternfly/components/Button/button.scss:
```diff
@@ -143,6 +143,12 @@
   --#{$button}--m-plain--active--Color: var(--#{$pf-global}--Color--100);
   --#{$button}--m-plain--disabled--Color: var(--#{$pf-global}--disabled-color--200);
   --#{$button}--m-plain--disabled--BackgroundColor: transparent;
+  --#{$button}--m-plain--m-inline--FontSize: inherit;
```
Ah interesting! 🤯  The codepen definitely helps show it

thatblindgeye commented:
> Use for a small/compact plain button

@mcoker that's something I had thought about, but it seemed less of an "inline" button and more of a "small" variant of a plain button. Maybe we offer both a plain+inline button as well as a small/compact+plain button variant, with the latter applying the smaller font size as well as a smaller padding? Throwing something together locally, here's what that could look like (keeping the same `var(--#{$button}--m-small--FontSize` font-size applied, then just updating the paddings so top/bottom gets the xs spacer and left/right gets sm, but could go a bit smaller all around):

![Normal plain button vs small plain button](https://github.com/patternfly/patternfly/assets/70952936/d048d10f-38c9-483f-b89c-ffc8a1588d17)

thatblindgeye commented on src/patternfly/components/Button/button.scss:
```diff
@@ -143,6 +143,12 @@
+  --#{$button}--m-plain--m-inline--PaddingRight: var(--#{$pf-global}--spacer--sm);;
+  --#{$button}--m-plain--m-inline--PaddingBottom: 0;
+  --#{$button}--m-plain--m-inline--PaddingLeft: var(--#{$pf-global}--spacer--sm);;
+  --#{$button}--m-plain--m-inline--LineHeight: normal;
```
Updating from `normal` to `var(--#{$pf-global}--LineHeight--md)` causes a slight difference in height between the button and whatever text it might be next to, with the button height slightly exceeding the height of the text next to it.

With the global var:

![Text height next to plain inline button with global var](https://github.com/patternfly/patternfly/assets/70952936/0f3bc61b-b5ae-4bd1-a3dc-db04ebbba050)
![Plain inline button height with global var](https://github.com/patternfly/patternfly/assets/70952936/d5f33810-2291-4364-b85d-e77405c5ed52)


With the `normal` keyword:

![Text height next to plain inline button with normal keyword](https://github.com/patternfly/patternfly/assets/70952936/0ec30cf9-3640-41cb-a648-8a2bb0ba94ef)
![Plain inline button with normal keyword](https://github.com/patternfly/patternfly/assets/70952936/8810a59d-da74-4f69-926e-eaabe4189c70)

Screenshots are from Firefox but Chrome has similar height differences, which get more significant as the font size increases. Though this could possibly depend how exactly a plain+inline button would actually be used... wdyt?

mcoker commented:
@thatblindgeye that all seems on point to me. We call those buttons "small" now. I wonder if this is a case for a "compact" button variant, assuming there could be a meaningful distinction between small and compact.

I know I created the original issue for this change, but I'm not totally clear on the role of an inline button. And I'm wondering if a compact button would negate any immediate need for an inline variant. From what I can tell, the padding/math in your inline screenshots looks good to me though.

I'm interested in the user's comment from the issue around their need for this feature, too, as well as any design input. cc @andrew-ronaldson @lboehling 

Side note, I'd love to see the "small" buttons in a separate example block called "small", they get lost in the variations block now.

thatblindgeye commented:
If we wanted to differentiate between "small" and "compact", "small" could refer to the font size while "compact" refer to both the font size *and* padding (or just padding?).

That all sounds good, though, will wait for some feedback from everyone else. Depending where we end up going with this PR I'll also move those small buttons to a separate example as part of this.

nicolethoen commented:
> @nicolethoen do you recall that conversation? I looked in the office hours agenda before that comment and skimmed a couple of the recordings and didn't see that discussion.

I don't - I looked at which office hours he'd be referring to - the one on March 15. And it must have been a question posed after the recording was stopped 🙁

andrew-ronaldson reviewed (approved):
Looks good to me but defer to Coker and Sarah on any outstanding code changes.

thatblindgeye commented:
@andrew-ronaldson @mcoker this can use another look after latest updates. I replaced the previous inline styling with a simpler compact styling as that was along the lines of the initial use-case. 

I also moved the small buttons to their own example per convo above and to align with the React examples.

mcoker reviewed (approved):
💪🚀👍

andrew-ronaldson commented:
Happy days. Looks great!

srambach commented on src/patternfly/components/Button/examples/Button.md:
```diff
@@ -406,6 +411,7 @@ Semantic buttons and links are important for usability as well as accessibility.
 | `.pf-m-warning` | `.pf-v5-c-button` | Modifies for warning styles. |
 | `.pf-m-link` | `.pf-v5-c-button` | Modifies for link styles. This button has no background or border and is styled as a link. This button would commonly appear in a form and may include an icon. |
 | `.pf-m-plain` | `.pf-v5-c-button` | Modifies for icon styles. This button has no background or border, uses a standard text color, and is used for `.pf-m-plain` icon buttons such as close, expand, kebab, etc. |
+| `.pf-m-no-padding` | `.pf-v5-c-button.pf-m-plain` | Modifies a plain button to remove padding. Should only be used when the button is inline within a sentence or block of text. Multiple plain buttons without padding should not be placed adjacent to one another without spacing between them. |
```
I'd make the second sentence a full sentence - "This modifier should only be used when the button is inline within a sentence or block of text."
And then maybe simplify the last sentence? I don't love this wording but was trying to make it what TO do rather than what not to do.
"Adjacent plain buttons without padding should always have spacing between them."

srambach reviewed (approved):
👍

patternfly-build commented:
:tada: This PR is included in version 5.1.0-prerelease.29 :tada:

The release is available on:
- [GitHub release](https://github.com/patternfly/patternfly/releases/tag/prerelease-v5.1.0-prerelease.29)
- [npm package (@prerelease dist-tag)](https://www.npmjs.com/package/@patternfly/patternfly/v/5.1.0-prerelease.29)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
